crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,767 @@
+    if (serial != NULL) {
+        ASN1_INTEGER_free(tmpl->serialNumber);
+        if ((tmpl->serialNumber = ASN1_INTEGER_dup(serial)) == NULL)
+            goto oom;
+    }
+    return 1;
+
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_CERTTEMPLATE_FILL, ERR_R_MALLOC_FAILURE);
+    return 0;
+}
+
+
+/*
+ * Decrypts the certificate in the given encryptedValue
+ * this is needed for the indirect PoP method as in RFC 4210 section 5.2.8.2
+ *
+ * returns a pointer to the decrypted certificate
+ * returns NULL on error or if no certificate available
+ */
+X509 *OSSL_CRMF_ENCRYPTEDVALUE_get1_encCert(OSSL_CRMF_ENCRYPTEDVALUE *ecert,
+                                            EVP_PKEY *pkey)
+{
+    X509 *cert = NULL; /* decrypted certificate */
+    EVP_CIPHER_CTX *evp_ctx = NULL; /* context for symmetric encryption */
+    unsigned char *ek = NULL; /* decrypted symmetric encryption key */
+    const EVP_CIPHER *cipher = NULL; /* used cipher */
+    unsigned char *iv = NULL; /* initial vector for symmetric encryption */
+    unsigned char *outbuf = NULL; /* decryption output buffer */
+    const unsigned char *p = NULL; /* needed for decoding ASN1 */
+    int symmAlg = 0; /* NIDs for symmetric algorithm */
+    int n, outlen = 0;
+    EVP_PKEY_CTX *pkctx = NULL; /* private key context */
+
+    if (ecert == NULL || ecert->symmAlg == NULL || ecert->encSymmKey == NULL
+        || ecert->encValue == NULL || pkey == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_ENCRYPTEDVALUE_GET1_ENCCERT,
+                CRMF_R_NULL_ARGUMENT);
+        return NULL;
+    }
+    if ((symmAlg = OBJ_obj2nid(ecert->symmAlg->algorithm)) == 0) {
+        CRMFerr(CRMF_F_OSSL_CRMF_ENCRYPTEDVALUE_GET1_ENCCERT,
+                CRMF_R_UNSUPPORTED_CIPHER);
+        return NULL;
+    }
+
+    /* first the symmetric key needs to be decrypted */
+    if ((pkctx = EVP_PKEY_CTX_new(pkey, NULL)) != NULL
+            && EVP_PKEY_decrypt_init(pkctx)) {
+        ASN1_BIT_STRING *encKey = ecert->encSymmKey;
+        size_t eksize = 0;
+
+        if (EVP_PKEY_decrypt(pkctx, NULL, &eksize, encKey->data, encKey->length)
+                <= 0
+            || (ek = OPENSSL_malloc(eksize)) == NULL
+            || EVP_PKEY_decrypt(pkctx, ek, &eksize, encKey->data,
+                                encKey->length) <= 0) {
+            CRMFerr(CRMF_F_OSSL_CRMF_ENCRYPTEDVALUE_GET1_ENCCERT,
+                    CRMF_R_ERROR_DECRYPTING_SYMMETRIC_KEY);
+            goto end;
+        }
+    } else {
+        goto oom;
+    }
+
+    /* select symmetric cipher based on algorithm given in message */
+    if ((cipher = EVP_get_cipherbynid(symmAlg)) == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_ENCRYPTEDVALUE_GET1_ENCCERT,
+                CRMF_R_UNSUPPORTED_CIPHER);
+        goto end;
+    }
+    if ((iv = OPENSSL_malloc(EVP_CIPHER_iv_length(cipher))) == NULL)
+        goto oom;
+    if (!ASN1_TYPE_get_octetstring(ecert->symmAlg->parameter, iv,
+                                   EVP_CIPHER_iv_length(cipher)))
```

Comment:
What happens if the iv length is less than expected? Won't it fail in the EVP_DecryptInit call?

Reply:
I will look into that

Reply:
@mpeylo, when are you going to look into this?

Reply:
If the IV retrieved from the ASN.1 is too short the data in _iv_ is likely to be wrong. Without having attempted to reproduce this in practice, I doubt that EVP_DecryptInit would fail as the memory size of _iv_ is appropriate, only with some 0-bytes at the end. However, it is likely that the decryption would fail in the end unless the content _iv_ corresponds to what was used on the peer side as IV for encyrypting.
So, in principle, such malformed IV could be explicitly caught as an explicit error with appropriate debug information - whether it is strictly needed I am not sure. However, I'd anyway touch that to put some debugging output before the "goto end" in line 728.